test/unit/extract/excel_test.py

```diff
@@ -115,3 +118,39 @@ def test_read_excel_calls(mock_read_excel):
 
     # TODO(rousik@gmail.com): need to figure out how to test process_$x methods.
     # TODO(rousik@gmail.com): we should test that empty columns are properly added.
+
+
+@pytest.mark.parametrize(
+    "dataset, expected_years",
+    (
+        ("eia860", set(range(2001, 2022))),
+        ("eia861", set(range(2001, 2022))),
+        ("eia923", set(range(2001, 2022))),
+    ),
+)
+def test_years_from_settings(dataset, expected_years):
+    years_from_settings = excel.years_from_settings_factory(dataset)
+
+    with build_op_context(
+        resources={"dataset_settings": DatasetsSettings()}
+    ) as context:
+        # Assert actual years are a superset of expected. Instead of doing
```

Comment:
Possibly, but I'm struggling to work out how to set years. The below didn't work, because the settings objects are immutable. I can't work out how to create them and pass years in though...

```diff
diff --git a/test/unit/extract/excel_test.py b/test/unit/extract/excel_test.py
index 767e1d04..1b2be0f7 100644
--- a/test/unit/extract/excel_test.py
+++ b/test/unit/extract/excel_test.py
@@ -123,16 +123,21 @@ class TestGenericExtractor(unittest.TestCase):
 @pytest.mark.parametrize(
     "dataset, expected_years",
     (
-        ("eia860", set(range(2001, 2022))),
-        ("eia861", set(range(2001, 2022))),
-        ("eia923", set(range(2001, 2022))),
+        ("eia860", [2001, 2002]),
+        ("eia861", [2003, 2004, 2005]),
+        ("eia923", [2022]),
     ),
 )
 def test_years_from_settings(dataset, expected_years):
     years_from_settings = excel.years_from_settings_factory(dataset)
 
+    settings = DatasetsSettings()
+    settings.eia.eia860.years = [2001, 2002]
+    settings.eia.eia861.years = [2003, 2004, 2005]
+    settings.eia.eia923.years = [2022]
+
     with build_op_context(
-        resources={"dataset_settings": DatasetsSettings()}
+        resources={"dataset_settings": settings}
     ) as context:
         # Assert actual years are a superset of expected. Instead of doing
         # an equality check, this avoids having to update expected years
```

Reply:
Oh yeah, this datasource metadata thing is pretty weird and complicated - lots of stuff that reads from disk by default. I think we can keep your workaround for now 😰